src/program.hpp

```diff
@@ -261,13 +273,32 @@ class spir_binary {
         m_image_metadata[name][ordinal].set_data_type(offset);
     }
 
+    void add_printf_descriptor(printf_descriptor&& desc) {
+        m_printf_descriptors[desc.printf_id] = desc;
+    }
+
+    void set_printf_buffer_info(const printf_buffer_desc_info& info) {
+        m_printf_buffer_info.reset(new printf_buffer_desc_info(info));
+    }
+
+    const printf_buffer_desc_info* printf_buffer_info() const {
+        return m_printf_buffer_info.get();
+    }
+
+    const std::unordered_map<uint32_t, printf_descriptor>
+    get_printf_descriptors() const {
+        return m_printf_descriptors;
+    }
+
 private:
     spv_context m_context;
     std::vector<uint32_t> m_code;
     std::vector<sampler_desc> m_literal_samplers;
     std::unordered_map<pushconstant, pushconstant_desc> m_push_constants;
     std::unordered_map<spec_constant, uint32_t> m_spec_constants;
     image_metadata_map m_image_metadata;
+    std::unordered_map<uint32_t, printf_descriptor> m_printf_descriptors;
+    std::unique_ptr<printf_buffer_desc_info> m_printf_buffer_info;
```

Comment:
This doesn't really need to be dynamically allocated, does it? It's not really a performance concern given the path it's happening on but I thought I'd ask in case I missed a good reason for it.

Reply:
It doesn't, I think my original thought here was that `m_printf_buffer_info` is optional so it may or may not be allocated for a program. I've updated this, now the buffer size defaults to 0 which is used to represent the absence of printf buffer info.